File: patterns/01-design-patterns/03-behavioral/11-interpreter/python/interpreter.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
def resolve(token,context):
    try:return int(token)
    except ValueError:
        if token not in context:raise PatternError("unknown_identifier")
        return context[token]
class Comparison:
    def __init__(self,left,operator,right):self.left=left;self.operator=operator;self.right=right
    def interpret(self,context,counter):
        counter[0]+=1;left,right=resolve(self.left,context),resolve(self.right,context)
        return left>=right if self.operator==">=" else left==right
class And:
    def __init__(self,children):self.children=children
    def interpret(self,context,counter):return all(child.interpret(context,counter) for child in self.children)
def parse(expression):
    tokens=expression.split();children=[];index=0
    while index<len(tokens):
        if index+2>=len(tokens):raise PatternError("invalid_expression")
        left,operator,right=tokens[index:index+3]
        if operator not in {">=","=="}:raise PatternError("unsupported_operator")
        children.append(Comparison(left,operator,right));index+=3
        if index<len(tokens):
            if tokens[index]!="and":raise PatternError("invalid_expression")
            index+=1
    if not children:raise PatternError("invalid_expression")
    return And(children)
def evaluate(input_data):
    counter=[0];allowed=parse(input_data["expression"]).interpret(input_data.get("context",{}),counter);return{"allowed":allowed,"comparisons":counter[0]}

File: patterns/01-design-patterns/03-behavioral/11-interpreter/go/interpreter.go
package interpreter

import "strconv"
import "strings"

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type Expression interface {
	Interpret(map[string]int, *int) (bool, error)
}
type Comparison struct{ left, operator, right string }

func resolve(token string, context map[string]int) (int, error) {
	if value, err := strconv.Atoi(token); err == nil {
		return value, nil
	}
	value, ok := context[token]
	if !ok {
		return 0, &PatternError{code: "unknown_identifier"}
	}
	return value, nil
}
func (c Comparison) Interpret(context map[string]int, counter *int) (bool, error) {
	*counter++
	left, err := resolve(c.left, context)
	if err != nil {
		return false, err
	}
	right, err := resolve(c.right, context)
	if err != nil {
		return false, err
	}
	if c.operator == ">=" {
		return left >= right, nil
	}
	return left == right, nil
}

type And struct{ children []Expression }

func (a And) Interpret(context map[string]int, counter *int) (bool, error) {
	for _, child := range a.children {
		ok, err := child.Interpret(context, counter)
		if err != nil {
			return false, err
		}
		if !ok {
			return false, nil
		}
	}
	return true, nil
}
func parseExpression(text string) (Expression, error) {
	tokens := strings.Fields(text)
	children := []Expression{}
	for index := 0; index < len(tokens); {
		if index+2 >= len(tokens) {
			return nil, &PatternError{code: "invalid_expression"}
		}
		operator := tokens[index+1]
		if operator != ">=" && operator != "==" {
			return nil, &PatternError{code: "unsupported_operator"}
		}
		children = append(children, Comparison{tokens[index], operator, tokens[index+2]})
		index += 3
		if index < len(tokens) {
			if tokens[index] != "and" {
				return nil, &PatternError{code: "invalid_expression"}
			}
			index++
		}
	}
	if len(children) == 0 {
		return nil, &PatternError{code: "invalid_expression"}
	}
	return And{children}, nil
}
func Evaluate(input map[string]any) (any, error) {
	expression, err := parseExpression(input["expression"].(string))
	if err != nil {
		return nil, err
	}
	context := map[string]int{}
	for key, value := range input["context"].(map[string]any) {
		context[key] = int(value.(float64))
	}
	count := 0
	allowed, err := expression.Interpret(context, &count)
	if err != nil {
		return nil, err
	}
	return map[string]any{"allowed": allowed, "comparisons": count}, nil
}
